pkg/ruler/grpcroundtripper.go

```diff
@@ -0,0 +1,94 @@
+package ruler
+
+import (
+	"bytes"
+	"io"
+	"net/http"
+	"net/url"
+
+	"github.com/go-kit/log"
+	"github.com/grafana/dskit/httpgrpc"
+	"github.com/grafana/dskit/middleware"
+	"github.com/grafana/mimir/pkg/util"
+	"github.com/prometheus/client_golang/prometheus"
+	"go.opentelemetry.io/contrib/instrumentation/google.golang.org/grpc/otelgrpc"
+	"google.golang.org/grpc"
+)
+
+// buffer exists to avoid extra copies of body data when adaping httpgrpc.HTTPClient to an http.RoundTripper
+type buffer struct {
+	buff []byte
+	io.ReadCloser
+}
+
+func wrapBuffer(b []byte) io.ReadCloser {
+	return &buffer{buff: b, ReadCloser: io.NopCloser(bytes.NewReader(b))}
+}
+
+func readAll(r io.Reader) ([]byte, error) {
+	if b, ok := r.(*buffer); ok {
```

Comment:
the approach the query-frontned package took was to retun a struct which implements `Bytes()`. I think that'll work a bit better once this is in the httpgrpcutil package